lldb/source/Target/SwiftLanguageRuntimeNames.cpp

```diff
@@ -131,9 +159,113 @@ static ThunkAction GetThunkAction(ThunkKind kind) {
+      return;
+    }
+
+    // TEMPORARY HACK WORKAROUND
+    auto fn_start = sc.symbol->GetFileAddress();
+    auto fn_end = sc.symbol->GetFileAddress() + sc.symbol->GetByteSize();
+    int line_entry_count = 0;
+    if (sc.comp_unit)
+      if (auto line_table = sc.comp_unit->GetLineTable())
+        for (int i = 0; i < line_table->GetSize(); ++i) {
+          LineEntry line_entry;
+          if (line_table->GetLineEntryAtIndex(i, line_entry)) {
+            auto line_start =
+                line_entry.range.GetBaseAddress().GetFileAddress();
+            if (line_start >= fn_start && line_start < fn_end) {
+              if (++line_entry_count >= 2)
+                // This is an async function with a valid body of code, no step
+                // into `swift_task_switch` required.
+                return;
+            }
+          }
+        }
+
+    assert(sc.function);
+    m_step_in_plan_sp = std::make_shared<ThreadPlanStepInRange>(
+        thread, sc.function->GetAddressRange(), sc, "swift_task_switch",
+        RunMode::eAllThreads, eLazyBoolNo, eLazyBoolNo);
+  }
+
+  void DidPush() override {
+    if (m_step_in_plan_sp)
+      PushPlan(m_step_in_plan_sp);
+  }
+
+  bool ValidatePlan(Stream *error) override { return (bool)m_step_in_plan_sp; }
+
+  void GetDescription(Stream *s, lldb::DescriptionLevel level) override {
+    // TODO: Implement completely.
+    s->PutCString("ThreadPlanStepInAsync");
+  }
+
+  // Composite thread plans never directly explain a stop.
+  bool DoPlanExplainsStop(Event *event) override { return false; }
+
+  // Async stops happen via breakpoint.
+  bool ShouldStop(Event *event) override { return false; }
+
+  bool MischiefManaged() override {
+    if (!m_step_in_plan_sp->IsPlanComplete()) {
+      return false;
+    }
+
+    if (!m_step_in_plan_sp->PlanSucceeded()) {
+      // If the step in fails, then this plan fails.
+      SetPlanComplete(false);
+      return true;
+    }
+
+    if (!m_async_breakpoint_sp) {
+      m_async_breakpoint_sp = CreateAsyncBreakpoint(GetThread());
+    }
+
+    SetPlanComplete();
+    return true;
+  }
+
+  // Override ShouldStop of previous step plan.
+  bool ShouldAutoContinue(Event *event) override { return true; }
+
+  bool WillStop() override { return false; }
+
+  lldb::StateType GetPlanRunState() override { return eStateRunning; }
+
+  bool StopOthers() override { return false; }
+
+private:
+  static BreakpointSP CreateAsyncBreakpoint(Thread &thread) {
+    auto frame_sp = thread.GetStackFrameAtIndex(0);
+    auto reg_ctx = frame_sp->GetRegisterContext();
+    auto task_reg = reg_ctx->ConvertRegisterKindToRegisterNumber(
+        RegisterKind::eRegisterKindGeneric, LLDB_REGNUM_GENERIC_ARG1);
+    auto task_ptr = reg_ctx->ReadRegisterAsUnsigned(task_reg, 0);
+    const auto fn_ptr_field = task_ptr + 0x28;
+    auto process_sp = thread.GetProcess();
+    Status error;
+    auto fn_ptr = process_sp->ReadPointerFromMemory(fn_ptr_field, error);
+    auto breakpoint =
```

Comment:
That's a shared pointer, so probably breakpoint_sp?

Reply:
done